### ISSUE TITLE
[HOLD for payment 2023-09-21] [$1000]  App allows task title update for completed task using deep link and throws error on update

### ISSUE THREAD

**Issue by kavimuru** [Expensify/App#22451]
If you haven’t already, check out our [contributing guidelines](https://github.com/Expensify/ReactNativeChat/blob/main/contributingGuides/CONTRIBUTING.md) for onboarding and email contributors@expensify.com to request to join our Slack channel!
___

## Action Performed:
1. Open the app
2. Click on plus and click on 'Assign a task'
3. Write any title and assign task to anyone
4. In task report, click on task to open task title edit
5. Copy the URL and send it to any other report
6. Mark the task as done
7. Click on link sent in step 5, edit the title and click on save
8. Observe that task report in LHN now has red dot for error
## Expected Result: 
Once task is completed, app should not allow to edit task title by deep link
## Actual Result: 
App allows to edit task title using deep link even after task is marked as completed and if edited using the link, app displays red dot in LHN on task report
## Workaround:
Can the user still use Expensify without this being fixed? Have you informed them of the workaround?

## Platforms:

Which of our officially supported platforms is this issue occurring on?
- [x] Android / native
- [x] Android / Chrome
- [ ] iOS / native
- [ ] iOS / Safari
- [x] MacOS / Chrome / Safari
- [ ] MacOS / Desktop

**Version Number:** 1.3.38-3
**Reproducible in staging?:** y
**Reproducible in production?:** y
**If this was caught during regression testing, add the test name, ID and link from TestRail:**
**Email or phone of affected tester (no customers):**
**Logs:** https://stackoverflow.com/c/expensify/questions/4856
**Notes/Photos/Videos:** Any additional supporting documentation

https://github.com/Expensify/App/assets/43996225/54d911ae-54de-4062-8083-3be2cb7aa2b2

https://github.com/Expensify/App/assets/43996225/e299c4b5-9b41-4c50-bffa-47e19f426929

**Expensify/Expensify Issue URL:**
**Issue reported by:** @dhanashree-sawant
**Slack conversation:** https://expensify.slack.com/archives/C049HHMV9SM/p1688742180722969

[View all open jobs on GitHub](https://github.com/Expensify/App/issues?q=is%3Aopen+is%3Aissue+label%3A%22Help+Wanted%22)

<details><summary>Upwork Automation - Do Not Edit</summary>
    <ul>
        <li>Upwork Job URL: https://www.upwork.com/jobs/~01df4c0b6543fe49a8</li>
        <li>Upwork Job ID: 1678539598526873600</li>
        <li>Last Price Increase: 2023-08-28</li>
<li>Automatic offers: </li>
<ul>
<li>BhuvaneshPatil | Contributor | 26456675</li>
<li>dhanashree-sawant | Reporter | 26456679</li>
</ul></li>
    </ul>
</details>

**Comment by BhuvaneshPatil** [comment]
## Proposal

### Please re-state the problem that we are trying to solve in this issue.
App allows task title update for completed task using deep link and throws error on update

### What is the root cause of that problem?
We are not having any check (whether task is open) when we are opening `TaskDescriptionPage` using link. That's causing this issue.

### What changes do you think we should make in order to solve the problem?
We can wrap the `TaskDescriptionPage` with `FullPageNotFoundView` component.
For `shouldShow` prop, we can add the required checks - 
- If the report is open or not
- If the current user has access to edit the description

This has advantage over `withReportOrNotFound` HOC. We can add custom message to show in two scenarios described above.
We can pass custom `subtitleKey` to `FullPageNotFoundView`.

### What alternative solutions did you explore? (Optional)
If we don't want to wrap the `FullPageNotFoundView`, we can use `withReportOrNotFound` here we can add additional condition to check if report is open or not.
```
report.stateNum === CONST.REPORT.STATE_NUM.OPEN && report.statusNum === CONST.REPORT.STATUS.OPEN;
```

By using this solution we can't provide detailed info to user why he/she can't access the page.

**Comment by twisterdotcom** [comment]
Wow, good find.

**Comment by twisterdotcom** [comment]
@eVoloshchak - any of these proposals look good?

**Comment by twisterdotcom** [comment]
Bump @eVoloshchak. I'll reassign to another C+ tomorrow if you're a bit busy, don't worry.

**Comment by eVoloshchak** [comment]
Thank you for the proposals, everyone!
The first four proposals are quite similar (they can't be that different, since all we need to do is conditionally display an error)

I think we should proceed with @Nodebrute's [proposal](https://github.com/Expensify/App/issues/22451#issuecomment-1634199434), it is adding an important check if user has permission to edit/view the task title/description in addition to checking if task is completed. 

We can use `FullPageNotFoundView`, there seem to be two cases:

1. The task is completed -> show `FullPageNotFoundView` with something like 'Can't edit task when it's completed'
2. User isn't Task Assignee, Task Assigner or Policy Admin (can't see or edit the task) ->  show `FullPageNotFoundView` with 'You can't access this task'

🎀👀🎀 C+ reviewed!

**Comment by BhuvaneshPatil** [comment]
@eVoloshchak @twisterdotcom 

https://github.com/Expensify/App/issues/21580 - This is issue that is being worked on right now. Please have a look.

**Comment by twisterdotcom** [comment]
Just waiting on your input @joelbettner!

**Comment by twisterdotcom** [comment]
@joelbettner bump on this please.

**Comment by BhuvaneshPatil** [comment]
@Nodebrute Have a look at this comment - https://github.com/Expensify/App/issues/21580#issuecomment-1614907480

It was approved then moved to internal.

**Comment by alitoshmatov** [comment]
## Proposal

### Please re-state the problem that we are trying to solve in this issue.
App allows task title update for completed task using deep link and throws error on update
### What is the root cause of that problem?
We have following routes:
<img width="491" alt="Screenshot 2023-07-25 at 11 01 18 AM" src="https://github.com/Expensify/App/assets/59907218/30bd62d9-e2ae-4f8d-bc48-cea76e22e50d">
We are not checking if user has access to this pages and can edit the task. Also when we manually enter this routes in non-task reports this pages show up and changes effect the report(at least locally) 
### What changes do you think we should make in order to solve the problem?

So we should add `FullPageNotFoundView` with following conditions:
Check if report is task report `!ReportUtils.isTaskReport(props.report)`
Check if report is completed `ReportUtils.isCompletedTaskReport(props.report)`
Check if user has permission to edit task `Task.isTaskAssigneeOrTaskOwner(props.report, props.session.accountID)`, `PolicyUtils.isPolicyAdmin(policy)`

In following places:
1. Task title page
https://github.com/Expensify/App/blob/a32fa95627b8a5a9a093c8e05e79dd2facf9f8a9/src/pages/tasks/TaskTitlePage.js#L69-L70
2. Task description page, we should also add since it doesn't have access to the current report
```javascript
report: {
            key: ({route}) => `${ONYXKEYS.COLLECTION.REPORT}${getReportID(route)}`,
        },
```
https://github.com/Expensify/App/blob/a32fa95627b8a5a9a093c8e05e79dd2facf9f8a9/src/pages/tasks/TaskDescriptionPage.js#L58-L59

3. Task assignee is a bit different condition should be `props.report && !ReportUtils.isTaskReport(props.report)` since it can be opened for new tasks and `props.report` might be empty
https://github.com/Expensify/App/blob/a656c82d0ebf1d3954575084555a004efc47035b/src/pages/tasks/TaskAssigneeSelectorModal.js#L183
Also we should add 
```javascript 
report: {
            key: ({route}) => `${ONYXKEYS.COLLECTION.REPORT}${getReportID(route)}`,
        },
```

### What alternative solutions did you explore? (Optional)

**Comment by alitoshmatov** [comment]
Coming from here https://github.com/Expensify/App/issues/23473 where we have similar problem with report pages. 

This issue is also present for non task reports:

> We are not checking if user has access to this pages and can edit the task. Also when we manually enter this routes in non-task reports this pages show up and changes effect the report(at least locally)


That's why posted proposal with solution to that also

**Comment by DanutGavrus** [comment]
@joelbettner 
With [this](https://github.com/Expensify/App/issues/22451#issuecomment-1634351931) implementation we can easily create rule sets which prevent user accessing sub-pages he shouldn't. We'll have them in just 1 place for the whole app instead of fixing each deeplink in its own component one by one. Seems to also make it easy to prevent/fix future similar issues. 

cc: @abdulrahuman5196 @thienlnam @stitesExpensify 
I'm curious to hear your thoughts about this approach as it seems different from other proposals. Maybe it's a good approach, maybe not. 🤔

**Comment by joelbettner** [comment]
> @eVoloshchak @joelbettner @twisterdotcom they are handling it from backend right now. It is marked as internal and no previous proposal has mentioned an effective way to handle it in frontend. We can deal with frontend in this issue.

Can you link to the issue in which the backend checks are being worked on? Are the backend changes complete? Will the front end changes be affected by whatever is written for the backend?

If the backend changes are complete and we are ready to start working on the front end changes, @eVoloshchak I'm not entirely sure which proposal I'm supposed to be reviewing since several other proposals were recently submitted.

Let me know.

**Comment by BhuvaneshPatil** [comment]
This is the issue that I've reported - https://github.com/Expensify/App/issues/21580.

For above issue, [proposal](https://github.com/Expensify/App/issues/21580#issuecomment-1607810113) was [accepted](https://github.com/Expensify/App/issues/21580#issuecomment-1614907480) by @abdulrahuman5196  there are lot of comments in that issue thread  on approach to follow. And the time when issue was reported and proposal was accepted we didn't have the `TaskView` that we have now. 

@thienlnam [decided](https://github.com/Expensify/App/issues/21580#issuecomment-1629396958) to work on this on backend. And issue was marked as internal.

The reason I added [this comment](https://github.com/Expensify/App/issues/22451#issuecomment-1650161799) is because while accepting proposal one of the reason was the accepted proposal has mentioned about adding other checks, and they were discussed way before and put on hold.

I have added the comment on that we shall wrap the components (that change title, description and assignee) with `FullPageNotFoundView` because many contributors are suggesting the same in their proposal.

**Comment by joelbettner** [comment]
> I don't think this frontend change will affect the backend./

As long as that is the case then I'm good with the proposal @eVoloshchak.

**Comment by thienlnam** [comment]
Adding the same comment as over [here](https://github.com/Expensify/App/issues/21580#issuecomment-1659476640), but let's wait until we update the permissions before moving forward with this change

**Comment by twisterdotcom** [comment]
Still held.

**Comment by twisterdotcom** [comment]
@thienlnam can we unhold this now?

**Comment by twisterdotcom** [comment]
@thienlnam can we unhold this now?

**Comment by thienlnam** [comment]
Yup, this is ready to go - let's see if it's still a problem

**Comment by twisterdotcom** [comment]
Apologies but I am finally going on Parental Leave. I am reassigning to another member of the team.

**Comment by BhuvaneshPatil** [comment]
best wishes @twisterdotcom

**Comment by BhuvaneshPatil** [comment]
@eVoloshchak Please have a look at this comment https://github.com/Expensify/App/issues/22451#issuecomment-1654671183. And the access check are extra work that is not needed. Because now anyone who can view the task can modify it. So please consider the base changes while selecting the proposal.

**Comment by eVoloshchak** [comment]
@BhuvaneshPatil, oh, I see, thank you for the explanation on this!
It became a bit convoluted with multiple issues dedicated to the same flow. 

While this issue was created for a specific case (accessing a task that is completed), we should cover all of the other cases where user doesn't have access to this page. This was first proposed by @BhuvaneshPatil in https://github.com/Expensify/App/issues/21580#issuecomment-1607810113, but that issue was closed (in favor of this one). I think it's fair to review proposals from both this issue and https://github.com/Expensify/App/issues/21580. 

This [proposal](https://github.com/Expensify/App/issues/21580#issuecomment-1607810113) from @BhuvaneshPatil was posted before this issue was even created, so I think we should proceed with that, since this issue was extended to implement all of the changes 

cc: @joelbettner

**Comment by eVoloshchak** [comment]
> This proposal doesn't have any update after the permissiom changed.

@dukenv0307, that is true, but I consider that a minor technical detail, the main idea of the proposal is the same

> We should use withReportOrNotFound HOC to control the loading when we access by deeplink before logging.

Same as above, we can achieve the same result using shouldShow prop for FullPageNotFoundView.
Additionally, as @BhuvaneshPatil  pointed out, that will allow us to show a custom message (i.e. "Can't edit a task that is completed/You don't have access to this task", etc)

**Comment by BhuvaneshPatil** [comment]
@eVoloshchak @joelbettner ^^ for futher process

**Comment by thienlnam** [comment]
@joelbettner I'll take this over since I'm already familiar

**Comment by BhuvaneshPatil** [comment]
Thank you @thienlnam .
Scope of this issue will be - 
- Disable updating title in case of completed task using deep link
- Add extra checks as discussed in #21580 

Please confirm.

**Comment by thienlnam** [comment]
Yup sounds good, thanks for sticking with these issues

**Comment by BhuvaneshPatil** [comment]
Hehe @thienlnam , I'll take it as a compliment

**Comment by dhanashree-sawant** [comment]
Hi @thienlnam, melvin bot has provided me with 50$ offer even though it is raised in july 😅. Can you check once you are available?

**Comment by sophiepintoraetz** [comment]
@dhanashree-sawant - I've paid you for $250.

**Comment by BhuvaneshPatil** [comment]
^^ @sophiepintoraetz @eVoloshchak @thienlnam 

Can we proceed further on this.
Also the PR was approved before 3 day time period, but was put on freeze.

**Comment by sophiepintoraetz** [comment]
Cool so:

Payouts due:

Issue Reporter: $250 done
Contributor: $1500 @BhuvaneshPatil 
Contributor+: $1500 @eVoloshchak (request a newdot payment once the BZ regression test checklist is done)

Eligible for 50% #urgency bonus? Y

**Comment by sophiepintoraetz** [comment]
@BhuvaneshPatil - for future reference, please wait until 7 days has passed before bumping - this has a weekly priority so while we endeavour to pay out on time, it's not always the case.

**Comment by BhuvaneshPatil** [comment]
thanks @sophiepintoraetz for mentioning.
I will keep that in mind from now.

**Comment by eVoloshchak** [comment]
- The PR that introduced the bug has been identified. Link to the PR: There isn't a PR that caused this, we've just added the logic that was missing. The whole task creation flow was implemented in https://github.com/Expensify/App/pull/17992, but at that point, this wasn't needed
- The offending PR has been commented on, pointing out the bug it caused and why, so the author and reviewers can learn from the mistake. Link to comment: N/A
- A discussion in #expensify-bugs has been started about whether any other steps should be taken (e.g. updating the PR review checklist) in order to catch this type of bug sooner. Link to discussion: I don't think an additional discussion is needed, we already have a checklist item to catch similar bugs: `If the PR modifies a component or page that can be accessed by a direct deeplink, I verified that the code functions as expected when the deeplink is used - from a logged in and logged out account`

**Comment by eVoloshchak** [comment]
**Regression Test Proposal**

1. Create a task
2. copy URL of task for description https://staging.new.expensify.com/r/<taskId>/description
3. Complete that task(mark as done)
4. Paste the URL into the browser
5. Verify that the user is presented with full-page not found view
6. Do the same steps (4-5) for title page - https://staging.new.expensify.com/r/<taskId>/title
7.  Do the same steps (4-5) for assignee page - https://staging.new.expensify.com/r/<taskId>/assignee

**Do we agree 👍 or 👎**

**Comment by eVoloshchak** [comment]
@sophiepintoraetz, please don't close an issue if the BZ checklist hasn't been completed yet
If the issue is closed, it isn't displayed in `Assigned Issues` section for C+, so there is no way to access it unless you specifically remember it exists (I stumbled upon this one randomly😅)
Thank you😊

**Comment by eVoloshchak** [comment]
Requested the payment via NewDot

**Comment by JmillsExpensify** [comment]
$1,500 payment approved for @eVoloshchak based on BZ summary.
